### PR TITLE
build(-system) fixes

### PR DESCRIPTION
**Pull request by paroj**
### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or another license that is incompatible with OpenCV
- [x] The PR is proposed to the proper branch
- [ ] There is a reference to the original bug report and related work
- [ ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ ] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Win64
build_image:Custom Win=msvs2019
allow_multiple_commits=1
```

### CODE REVIEW

File: cmake/OpenCVPCHSupport.cmake [hunk 1]
```diff
@@ -406,7 +406,9 @@ ENDMACRO(ADD_NATIVE_PRECOMPILED_HEADER)
 
 macro(ocv_add_precompiled_header_to_target the_target pch_header)
   if(PCHSupport_FOUND AND ENABLE_PRECOMPILED_HEADERS AND EXISTS "${pch_header}")
-    if(CMAKE_GENERATOR MATCHES "^Visual" OR CMAKE_GENERATOR MATCHES Xcode)
+    if(NOT CMAKE_VERSION VERSION_LESS "3.16")
+      target_precompile_headers(${the_target} PRIVATE ${pch_header})
```
Comment (alalek) [inline cmake/OpenCVPCHSupport.cmake:410]: Fix is not complete: https://pullrequest.opencv.org/buildbot/builders/precommit_windows64/builds/100106
(need to adapt, GET_NATIVE_PRECOMPILED_HEADER macro)

Also please add pre-check to bypass this flow: `AND NOT OPENCV_SKIP_CMAKE_BUILTIN_PCH`

**Comment by alalek** [comment]
There are linker failures from MSVS 2015 configuration:

```
test_intrin128.sse2.obj : error LNK2005: "void __cdecl opencv_test::hal::intrin128::cpu_baseline::test_hal_intrin_float16(void)" (?test_hal_intrin_float16@cpu_baseline@intrin128@hal@opencv_test@@YAXXZ) already defined in test_intrin.obj [C:\build\precommit_windows64\build\modules\core\opencv_test_core.vcxproj]
test_intrin128.sse2.obj : error LNK2005: "void __cdecl opencv_test::hal::intrin128::cpu_baseline::test_hal_intrin_float32(void)" (?test_hal_intrin_float32@cpu_baseline@intrin128@hal@opencv_test@@YAXXZ) already defined in test_intrin.obj [C:\build\precommit_windows64\build\modules\core\opencv_test_core.vcxproj]
...
```

**Comment by paroj** [comment]
it seems the PCH change unmasks some deeper issues regarding the tests.

I would suggest to just revert to the old PCH code for MSVC, as I lack the insight to fix the tests right now.

**Comment by alalek** [comment]
Looks like in this case compilation with PCH ignores some compilation flags (PCH file can't handle multiple sets of them).
I prepared a fix here: #22600 (to 3.4 branch, so need to wait for "Merge 3.4")
Please drop the last commit with "inline" changes.

Patch is validated here: https://pullrequest.opencv.org/buildbot/builders/precommit_windows64/builds/100140

**Review by alalek — APPROVED** [review]
Thank you :+1:

**Pull request merged**